tornado/http1connection.py

```diff
@@ -244,34 +248,36 @@ def _read_message(self, delegate):
                                 self.stream.close()
                                 raise ReturnFalse
             self._read_finished = True
-            if not self._write_finished or self.is_client:
+            if not self._write_finished or is_client:
                 need_delegate_close = False
                 with _ExceptionLoggingContext(app_log):
                     delegate.finish()
             # If we're waiting for the application to produce an asynchronous
             # response, and we're not detached, register a close callback
             # on the stream (we didn't need one while we were reading)
-            if (not self._finish_future.done() and
-                    self.stream is not None and
-                    not self.stream.closed()):
-                self.stream.set_close_callback(self._on_connection_close)
-                yield self._finish_future
-            if self.is_client and self._disconnect_on_finish:
+            finish_future = self._finish_future
+            if not finish_future.done():
+                stream = self.stream
+                if stream is not None and not stream.closed():
+                    stream.set_close_callback(self._on_connection_close)
+                    stream = None
```

Comment:
Why do you set the stream to `None` if this is going to be done on line 289?

Reply:
To release the reference. You mean 281. That's after the yield, which can be an eternity later.